shared-module/audiofilters/Distortion.c

```diff
@@ -0,0 +1,409 @@
+    uint32_t buffer_size, uint8_t bits_per_sample, bool samples_signed,
+    uint8_t channel_count, uint32_t sample_rate) {
+
+    // Basic settings every effect and audio sample has
+    // These are the effects values, not the source sample(s)
+    self->bits_per_sample = bits_per_sample; // Most common is 16, but 8 is also supported in many places
+    self->samples_signed = samples_signed; // Are the samples we provide signed (common is true)
+    self->channel_count = channel_count; // Channels can be 1 for mono or 2 for stereo
+    self->sample_rate = sample_rate; // Sample rate for the effect, this generally needs to match all audio objects
+
+    // To smooth things out as CircuitPython is doing other tasks most audio objects have a buffer
+    // A double buffer is set up here so the audio output can use DMA on buffer 1 while we
+    // write to and create buffer 2.
+    // This buffer is what is passed to the audio component that plays the effect.
+    // Samples are set sequentially. For stereo audio they are passed L/R/L/R/...
+    self->buffer_len = buffer_size; // in bytes
+
+    self->buffer[0] = m_malloc(self->buffer_len);
+    if (self->buffer[0] == NULL) {
+        common_hal_audiofilters_distortion_deinit(self);
+        m_malloc_fail(self->buffer_len);
+    }
+    memset(self->buffer[0], 0, self->buffer_len);
+
+    self->buffer[1] = m_malloc(self->buffer_len);
+    if (self->buffer[1] == NULL) {
+        common_hal_audiofilters_distortion_deinit(self);
+        m_malloc_fail(self->buffer_len);
+    }
+    memset(self->buffer[1], 0, self->buffer_len);
+
+    self->last_buf_idx = 1; // Which buffer to use first, toggle between 0 and 1
+
+    // Initialize other values most effects will need.
+    self->sample = NULL; // The current playing sample
+    self->sample_remaining_buffer = NULL; // Pointer to the start of the sample buffer we have not played
+    self->sample_buffer_length = 0; // How many samples do we have left to play (these may be 16 bit!)
+    self->loop = false; // When the sample is done do we loop to the start again or stop (e.g. in a wav file)
+    self->more_data = false; // Is there still more data to read from the sample or did we finish
+
+    // The below section sets up the effect's starting values.
+
+    synthio_block_assign_slot(drive, &self->drive, MP_QSTR_drive);
+    synthio_block_assign_slot(pre_gain, &self->pre_gain, MP_QSTR_pre_gain);
+    synthio_block_assign_slot(post_gain, &self->post_gain, MP_QSTR_post_gain);
+    synthio_block_assign_slot(mix, &self->mix, MP_QSTR_mix);
+
+    self->mode = mode;
+    self->soft_clip = soft_clip;
+}
+
+bool common_hal_audiofilters_distortion_deinited(audiofilters_distortion_obj_t *self) {
+    if (self->buffer[0] == NULL) {
+        return true;
+    }
+    return false;
+}
+
+void common_hal_audiofilters_distortion_deinit(audiofilters_distortion_obj_t *self) {
+    if (common_hal_audiofilters_distortion_deinited(self)) {
+        return;
+    }
```

Comment:
this check is not needed, because it's just fine to set the fields to NULL a second time. Or leave it in, it's not doing much besides costing a few bytes of flash.